smallrye-reactive-messaging-mqtt/src/main/java/io/smallrye/reactive/messaging/mqtt/session/impl/MqttClientSessionImpl.java

```diff
@@ -0,0 +1,672 @@
+import io.smallrye.reactive.messaging.mqtt.session.SessionState;
+import io.smallrye.reactive.messaging.mqtt.session.SubscriptionEvent;
+import io.smallrye.reactive.messaging.mqtt.session.SubscriptionState;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxException;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.impl.VertxInternal;
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
+import io.vertx.mqtt.MqttClient;
+import io.vertx.mqtt.messages.MqttConnAckMessage;
+import io.vertx.mqtt.messages.MqttPublishMessage;
+import io.vertx.mqtt.messages.MqttSubAckMessage;
+
+public class MqttClientSessionImpl implements MqttClientSession {
+
+    private static final Logger log = LoggerFactory.getLogger(MqttClientSessionImpl.class);
+
+    private final VertxInternal vertx;
+    private final MqttClientSessionOptions options;
+
+    // record the subscriptions
+    private final Map<String, RequestedQoS> subscriptions = new HashMap<>();
+    // record the pending subscribes
+    private final Map<Integer, LinkedHashMap<String, RequestedQoS>> pendingSubscribes = new HashMap<>();
+    // record the pending unsubscribes
+    private final Map<Integer, List<String>> pendingUnsubscribes = new HashMap<>();
+    // the provider for the reconnect delay
+    private final ReconnectDelayProvider reconnectDelay;
+
+    // the current state
+    private volatile SessionState state = SessionState.DISCONNECTED;
+    // drives to connection either to CONNECTED or DISCONNECTED
+    private volatile boolean running;
+    // subscription states
+    private final Map<String, SubscriptionState> subscriptionStates = new ConcurrentHashMap<>();
+
+    // holds the actual MQTT client connection
+    private MqttClient client;
+    // an optional reconnect timer
+    private Long reconnectTimer;
+
+    private volatile Handler<MqttPublishMessage> messageHandler;
+    private volatile Handler<Throwable> exceptionHandler;
+    private volatile Handler<SessionEvent> sessionStateHandler;
+    private volatile Handler<SubscriptionEvent> subscriptionStateHandler;
+    private volatile Handler<Integer> publishCompleteHandler;
+    private volatile Handler<Integer> publishCompletionExpirationHandler;
+    private volatile Handler<Integer> publishCompletionUnknownPacketIdHandler;
+
+    /**
+     * Create a new instance, which is not started.
+     *
+     * @param vertx The vert.x instance to use.
+     * @param options The client session options.
+     */
+    public MqttClientSessionImpl(final Vertx vertx, final MqttClientSessionOptions options) {
+        this.vertx = (VertxInternal) vertx;
+        this.options = options;
+        this.reconnectDelay = options.getReconnectDelay().createProvider();
+    }
+
+    @Override
+    public void start() {
+        this.vertx.runOnContext(x -> doStart());
```

Comment:
The start and stop should return Uni or Future and complete/fail when the action completes/fails

Reply:
There is no real "end" of the operation. That just starts the state machine driving the connection towards "CONNECTED".